event-web-framework/draft.markdown

```diff
@@ -0,0 +1,941 @@
+
+```lisp
+(defclass response ()
+  ((content-type
+    :accessor content-type :initform "text/html" :initarg :content-type)
+   (charset
+    :accessor charset :initform "utf-8")
+   (response-code
+    :accessor response-code :initform "200 OK" :initarg :response-code)
+   (keep-alive?
+    :accessor keep-alive? :initform nil :initarg :keep-alive?)
+   (body
+    :accessor body :initform nil :initarg :body)))
+```
+
+The second is an [SSE message](http://www.w3.org/TR/eventsource/), which we will use to send an incremental update to our clients.
+
+```lisp
+(defclass sse ()
+  ((id :reader id :initarg :id :initform nil)
+   (event :reader event :initarg :event :initform nil)
+   (retry :reader retry :initarg :retry :initform nil)
+   (data :reader data :initarg :data)))
+```
+
+We'll send an HTTP response whenever we receive a full HTTP request; however, how do we know when and where to send SSE messages without an originating client request?
+
+A simple solution is to register _channels_[^defparameter], to which we'll subscribe `socket`s as necessary.
+
+```lisp
+(defparameter *channels* (make-hash-table))
+
+(defmethod subscribe! ((channel symbol) (sock usocket))
+  (push sock (gethash channel *channels*))
+  nil)
+```
+
+[^defparameter]: We're incidentally introducing some new syntax here. This is our way of declaring a mutable variable. It has the form `(defparameter <name> <value> <optional docstring>)`.
+
+We can then `publish!` notifications to said channels as soon as they become available.
+
+```lisp
+(defmethod publish! ((channel symbol) (message string))
+  (awhen (gethash channel *channels*)
+	 (setf (gethash channel *channels*)
+	       (loop with msg = (make-instance 'sse :data message)
+		  for sock in it
+		  when (ignore-errors
+			 (write! msg sock)
+			 (force-output (socket-stream sock))
+			 sock)
+		  collect it))))
+```
+
+In `publish!`, we call `write!` to actually write an `sse` to a socket. We'll also need a specialization of `write!` on `response`s to write full HTTP responses as well. Let's handle the HTTP case first.
+
+```lisp
+(defmethod write! ((res response) (socket usocket))
+  (handler-case
+      (with-timeout (.2)
+	(let ((stream (flex-stream socket)))
+	  (flet ((write-ln (&rest sequences)
+		   (mapc (lambda (seq) (write-sequence seq stream)) sequences)
+		   (crlf stream)))
+	    (write-ln "HTTP/1.1 " (response-code res))
+	    (write-ln
+	     "Content-Type: " (content-type res) "; charset=" (charset res))
+	    (write-ln "Cache-Control: no-cache, no-store, must-revalidate")
+	    (when (keep-alive? res)
+	      (write-ln "Connection: keep-alive")
+	      (write-ln "Expires: Thu, 01 Jan 1970 00:00:01 GMT"))
+	    (awhen (body res)
+	      (write-ln "Content-Length: " (write-to-string (length it)))
+	      (crlf stream)
+	      (write-ln it))
+	    (values))))
+    (trivial-timeout:timeout-error ()
```

Comment:
Nope. If there's an error during a request `write!` we throw that socket out (we try to write an `error` response first, but at that point we just drop the client). Not entirely sure that's the *right* thing, but I'm not sure what else we could do if retries have made it up past TCP level.